### DISCUSSION TITLE
Insert an invocation expression after certain nodes

### DISCUSSION THREAD

**Discussion by eirannejad** [dotnet/roslyn#62139] [General]
Hello!

I'm trying to insert an invocation expression after certain nodes in my C# syntax tree using Roslyn but am running into an issue [similar to this](https://github.com/dotnet/roslyn/issues/57494)

Would any of you mind showing me how I can rewrite a syntax tree by inserting nodes (using `InsertNodeAfter`) in a `CSharpSyntaxRewriter`? Let's assume I'm trying to override `AssignmentExpression` and add an `InvocationExpression` right after the assignment

**Comment by CyrusNajmabadi** [comment]
Can you show an example of what the code looks like at the start, and what you want it to look like sheet the rewrite?  Thanks!

**Reply by eirannejad** [reply]
Similar to this...

**From:**
```
char key = Console.ReadKey();
if(key == 'A')
{
Console.WriteLine("You pressed A");
}
else
{
Console.WriteLine("You didn't press A");
}
```

**To:** (Added `LogConditionWasTrue();` lines)
```
char key = Console.ReadKey();
if(key == 'A')
{
LogConditionWasTrue();
Console.WriteLine("You pressed A");
}
else
{
Console.WriteLine("You didn't press A");
LogConditionWasFalse();
}
```

[This article here](https://joshvarty.com/2015/08/18/learn-roslyn-now-part-12-the-documenteditor/) explains how to do this with a `DocumentEditor` but not using a `CSharpSyntaxRewriter`

**Reply by PathogenDavid** [reply]
@eirannejad You seem to have accidentally posted the same code snippet twice.

**Reply by CyrusNajmabadi** [reply]
In this case, you'd need to rewrite the block Syntax for:

```
{
Console.WriteLine("You didn't press A");
}
```

You'd replace it with a block with two child statements instead of one.

**Reply by eirannejad** [reply]
@PathogenDavid @CyrusNajmabadi 😅 The actual example doesn't matter; and I know I need to replace a single node with two nodes. Going back to the original question, I'm getting an exception when `InsertNodeAfter` or even `Replace` a node with a node of a different type (e.g. `UsingDeclaration` with a `ExpressionStatement`). I'd love to get an snippet from anyone here that would show how a `CSharpSyntaxRewriter` would make the change shown above without throwing exceptions

**Reply by CyrusNajmabadi** [reply]
@eirannejad not at computer.  But when you hit the BlockSyntax, just replace it's Statement nodes list with the new list you build.  If youw ant to use InsertNodeAfter you'd need to to insert it after the ExpressionStatement not the expression.
